data-plane/core/src/test/java/dev/knative/eventing/kafka/broker/core/reconciler/impl/ResourcesReconcilerImplTest.java

```diff
@@ -0,0 +1,209 @@
+package dev.knative.eventing.kafka.broker.core.reconciler.impl;
+
+import static dev.knative.eventing.kafka.broker.contract.DataPlaneContract.Egress;
+import static dev.knative.eventing.kafka.broker.contract.DataPlaneContract.Filter;
+import static dev.knative.eventing.kafka.broker.contract.DataPlaneContract.Resource;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import dev.knative.eventing.kafka.broker.contract.DataPlaneContract;
+import java.util.Collections;
+import java.util.List;
+import org.junit.jupiter.api.Test;
+
+class ResourcesReconcilerImplTest {
+
+  @Test
+  void nullPointerExceptionWhenNoListenerIsProvided() {
+    assertThatThrownBy(() -> ResourcesReconcilerImpl.builder().build())
+      .isInstanceOf(NullPointerException.class);
+  }
+
+  @Test
+  void reconcileIngressAddNewResourcesWithoutIngress() {
+    new ResourceReconcilerTestRunner()
+      .enableIngressListener()
+      .step(List.of(
+        baseResource("1-1234").build(),
+        baseResource("2-1234").build()
+      ))
+      .done()
+      .run();
+  }
+
+  @Test
+  void reconcileIngressAddNewIngressAtSecondStep() {
+    new ResourceReconcilerTestRunner()
+      .enableIngressListener()
+      .step(List.of(
+        baseResource("1-1234")
+          .setIngress(DataPlaneContract.Ingress.newBuilder().setPath("/hello"))
+          .build(),
+        baseResource("2-1234")
+          .build()
+      ))
+      .newIngress("1-1234")
+      .done()
```

Comment:
I don't like `given` in this context honestly, for me `step` is more clear because it gives me the idea of a flow of subsequential things to do. `given` doesn't imply any sequence of events to succeed.
Regarding `expect`, this is not really needed (unless we want to add it as noop and use it for clarity)

Reply:
> I don't like given in this context honestly, for me step is more clear because it gives me the idea of a flow of subsequential things to do. given doesn't imply any sequence of events to succeed.

What about `enqueue` or simply `reconcile`? `step` seems too generic, and it isn't clear what it does.

> Regarding expect, this is not really needed (unless we want to add it as noop and use it for clarity)

Yes, for clarity.

Reply:
`reconcile` sounds good.

> Yes, for clarity.

With `reconcile` it seems clear enough to me... Maybe we need to change `done`?

```
      .reconcile(List.of(
        baseResource("1-1234")
          .setIngress(DataPlaneContract.Ingress.newBuilder().setPath("/hello"))
          .build(),
        baseResource("2-1234")
          .build()
      ))
      .newIngress("1-1234")
      .done()
```

Reply:
> Maybe we need to change done?

With what? `and()`? :thinking: :arrow_down: 

```java
      .reconcile(List.of(
        baseResource("1-1234")
          .addEgresses(egress("aaa"))
          .addEgresses(egress("bbb"))
          .addEgresses(egress("ccc"))
          .build()
      ))
      .expect()
      .newEgress("aaa")
      .newEgress("bbb")
      .newEgress("ccc")
      .and()
      .reconcile(List.of(
        baseResource("1-1234")
          .addEgresses(egress("aaa"))
          .addEgresses(baseEgress("bbb").setFilter(Filter.newBuilder().putAttributes("id", "hello")))
          .addEgresses(egress("ccc"))
          .build()
      ))
      .expect()
      .updatedEgress("bbb")
```

Reply:
Ok I tried to add expect, look nicer :smile:

Reply:
or `then()` instead of `and()`.

```java
      .reconcile(List.of(
        baseResource("1-1234")
          .addEgresses(egress("aaa"))
          .addEgresses(egress("bbb"))
          .addEgresses(egress("ccc"))
          .build()
      ))
      .expect()
      .newEgress("aaa")
      .newEgress("bbb")
      .newEgress("ccc")
      .then()
      .reconcile(List.of(
        baseResource("1-1234")
          .addEgresses(egress("aaa"))
          .addEgresses(baseEgress("bbb").setFilter(Filter.newBuilder().putAttributes("id", "hello")))
          .addEgresses(egress("ccc"))
          .build()
      ))
      .expect()
      .updatedEgress("bbb")
      .run()
```

Reply:
> With what? and()?

Nah because it may be confused as boolean operations... `next`?

Reply:
+1 for `then`.